airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/destination_async/FlushWorkers.java

```diff
@@ -53,93 +54,145 @@ public class FlushWorkers implements AutoCloseable {
-    log.info("Workers shut status: {}", workersShut);
+    try {
+
+      log.info("Closing flush workers -- waiting for all buffers to flush");
+      isClosing.set(true);
+      // wait for all buffers to be flushed.
+      while (true) {
+        final boolean anyRecordsLeft = bufferDequeue.getBufferedStreams()
+            .stream()
+            .anyMatch(streamDesc -> bufferDequeue.getQueueSizeInRecords(streamDesc).map(size -> size > 0).orElseThrow());
+        log.info("anyRecordsLeft: {}", anyRecordsLeft);
+        if (!anyRecordsLeft) {
+          break;
+        }
 
-    debugLoop.shutdownNow();
+        log.info("Waiting for all streams to flush.");
+        Thread.sleep(1000);
+      }
+      log.info("Closing flush workers -- all buffers flushed");
+
+      supervisorThread.shutdown();
+      final var supervisorShut = supervisorThread.awaitTermination(5L, TimeUnit.MINUTES);
+      log.info("Closing flush workers -- Supervisor shut status: {}", supervisorShut);
+
+      log.info("Closing flush workers -- Starting worker pool shutdown..");
+      workerPool.shutdown();
+      final var workersShut = workerPool.awaitTermination(5L, TimeUnit.MINUTES);
+      log.info("Closing flush workers -- Workers shut status: {}", workersShut);
+
+      debugLoop.shutdownNow();
+    } catch (final Exception e) {
+      log.error("Flush worker error: ", e);
+      flushFailure.propagateException(e);
+      throw new RuntimeException(e);
+    }
   }
 
   private void retrieveWork() {
-    // todo (cgardens) - i'm not convinced this makes sense. as we get close to the limit, we should
-    // flush more eagerly, but "flush all" is never a particularly useful thing in this world.
-    // if the total size is > n, flush all buffers
-    if (bufferDequeue.getTotalGlobalQueueSizeBytes() > BufferManager.TOTAL_QUEUES_MAX_SIZE_LIMIT_BYTES) {
-      flushAll();
-      return;
+    try {
+      log.info("retrieving queues to process");
+      final ThreadPoolExecutor threadPoolExecutor = (ThreadPoolExecutor) workerPool;
+      int allocatableThreads = threadPoolExecutor.getMaximumPoolSize() - threadPoolExecutor.getActiveCount();
+
+      while (allocatableThreads > 0) {
+        // when we are closing, flush regardless of how few items are in the queue.
+        final long computedQueueThreshold = isClosing.get() ? 0 : QUEUE_FLUSH_THRESHOLD_BYTES;
+
+        final Optional<StreamDescriptor> next = getNext(computedQueueThreshold);
+
+        if (next.isPresent()) {
+          final StreamDescriptor desc = next.get();
+          if (streamToInProgressWorkers.containsKey(desc)) {
+            streamToInProgressWorkers.get(desc).getAndAdd(1);
+          } else {
+            streamToInProgressWorkers.put(desc, new AtomicInteger(1));
+          }
+          allocatableThreads--;
+          flush(desc);
+        } else {
+          break;
+        }
+      }
+    } catch (final Exception e) {
```

Comment:
realized that we need to catch and propagate exceptions here too, because it also runs in a thread. i was confused for the longest time because there was an NPE, but it failed silently until I propagated the exception.

Reply:
you are right. I was focused on the thread pool and forgot about the scheduler thread - good catch